shell/platform/windows/direct_manipulation.cc

```diff
@@ -53,16 +56,31 @@ HRESULT DirectManipulationEventHandler::OnViewportStatusChanged(
             (int32_t) reinterpret_cast<int64_t>(this));
       }
     }
-  } else if (previous == DIRECTMANIPULATION_RUNNING) {
+  }
+  if (previous == DIRECTMANIPULATION_RUNNING) {
+    // Reset deltas to ensure only inertia values will be compared later.
+    last_pan_delta_x_ = 0.0;
+    last_pan_delta_y_ = 0.0;
     if (owner_->binding_handler_delegate) {
       owner_->binding_handler_delegate->OnPointerPanZoomEnd(
           (int32_t) reinterpret_cast<int64_t>(this));
     }
+  } else if (previous == DIRECTMANIPULATION_INERTIA) {
```

Comment:
We only compare the previous status but not the current, does it mean if the inertia movement dies out naturally it will also dispatch a cancel event?

Reply:
Current should always be DIRECTMANIPULATION_READY whether it dies out naturally or is cancelled. We can only tell if it's cancelled prematurely by seeing if the last event still had meaningful velocity left (hadn't already decelerated to near-zero).